File: Objective C/testTek/SFTableViewLayout.m
//
//  SFTableViewLayout.m
//  testTek
//
//  Created by Nicolas Bellon on 09/09/2015.
//  Copyright © 2015 stupeflix. All rights reserved.
//

#import "SFTableViewLayout.h"
#import "SFTableViewCell.h"

const NSString  *kAllConstraints    =  @"kAllConstraints";
const NSString  *kTopConstraint     =  @"kTopConstraint";
const NSString  *kBottomConstraint  =  @"kBottomConstraint";

@interface SFTableViewCell()

@property (nonatomic, weak) SFTableViewCell *previousCell;
@property (nonatomic, weak) SFTableViewCell *nextCell;

@end

@implementation SFTableViewLayout


/*  Dictionnaire de contraites de placement d'une cell dans la scrollview.
    En fonction de ses pointeurs previousCell et nextCell ses contraintes seront differentes
 */
+ (NSDictionary *)constraintWithCell:(SFTableViewCell *)pCell
{
    if (pCell == nil) return nil;
    
    NSMutableArray *dest =[NSMutableArray arrayWithCapacity:0];
    NSMutableDictionary *destDico = [NSMutableDictionary dictionaryWithCapacity:0];
    
    NSArray *verticalConstraints = nil;
    NSArray *cellHeight = nil;
    NSArray *horizontalConstraints = nil;
    NSArray *lastCellConstraints = nil;
    
    horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"[Cell(==heightWidth)]" options:0 metrics:@{@"heightWidth":[NSString stringWithFormat:@"%f", [SFTableViewCell width]]}
                                                                      views:@{@"Cell":pCell}];
    
    cellHeight = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[Cell(==heightSize)]" options:0 metrics:@{@"heightSize":[NSString stringWithFormat:@"%f", [SFTableViewCell height]]} views:@{@"Cell":pCell}];
    
    if (pCell.previousCell != nil)
    {
        verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[PreviousCell]-spaceBetweenCells-[Cell]" options:0 metrics:@{@"spaceBetweenCells":@"10"} views:@{@"Cell":pCell, @"PreviousCell":pCell.previousCell}];
    }
    else
    {
        verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[Cell]" options:0 metrics:0 views:@{@"Cell":pCell}];
    }
    
    if (pCell.nextCell == nil)
    {
        lastCellConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[Cell]-0-|" options:0 metrics:0 views:@{@"Cell":pCell}];
    }
    
    [dest addObjectsFromArray:horizontalConstraints];
    [dest addObjectsFromArray:cellHeight];
    [dest addObjectsFromArray:verticalConstraints];
    
    if (lastCellConstraints)
    {
        [dest addObjectsFromArray:lastCellConstraints];
    }
    [destDico setObject:dest.copy forKey:kAllConstraints];
    [destDico setObject:verticalConstraints.lastObject forKey:kTopConstraint];
    
    if (lastCellConstraints)
    {
        [destDico setObject:lastCellConstraints.lastObject forKey:kBottomConstraint];
    }
    
    return destDico;
}


/*  Dictionnaire de contraintes pour le placement de la vue en cours de mouvement
    Lors de son déplacement pendant la gesture, la cellule reçoit un nouveau set de contraintes
    qui la lie à la scroll view depuis le haut de la vue.
    C'est la constante de cette contrainte qui sera modifiée pour suivre le doigt
 */
+ (NSDictionary *)moovingConstraintWithCell:(SFTableViewCell *)pCell
{
    NSArray *horizontalConstraints = nil;
    NSArray *cellHeight = nil;
    NSArray *verticalConstraints = nil;
    NSMutableArray *allConstraints = [NSMutableArray arrayWithCapacity:0];
    
    NSMutableDictionary *dico = [NSMutableDictionary dictionaryWithCapacity:0];
    
    horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"[Cell(==heightWidth)]" options:0 metrics:@{@"heightWidth":[NSString stringWithFormat:@"%f", [SFTableViewCell width]]} views:@{@"Cell":pCell}];
    
    cellHeight = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[Cell(==heightSize)]" options:0 metrics:@{@"heightSize":[NSString stringWithFormat:@"%f", [SFTableViewCell height]]} views:@{@"Cell":pCell}];
    
    verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[Cell]" options:0 metrics:0 views:@{@"Cell":pCell}];
    
    [allConstraints addObjectsFromArray:horizontalConstraints];
    [allConstraints addObjectsFromArray:cellHeight];
    [allConstraints addObjectsFromArray:verticalConstraints];
    
    dico[kAllConstraints] = allConstraints.copy;
    dico[kTopConstraint] = verticalConstraints.lastObject;
    
    return dico;
}


/* Array de contraintes pour le placement de la scroll view */
+ (NSArray *)scrollConstraintsWithScrollView:(UIScrollView *)pScrollView
{
    NSMutableArray *dest = [NSMutableArray arrayWithCapacity:0];
    
    NSArray *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"|-0-[scrollView]-0-|" options:0 metrics:0 views:@{@"scrollView":pScrollView}];
    
    NSArray *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[scrollView]-0-|" options:0 metrics:0 views:@{@"scrollView":pScrollView}];
    
    [dest addObjectsFromArray:horizontalConstraints];
    [dest addObjectsFromArray:verticalConstraints];
    
    return dest;
}


@end

File: Swift/testTekSwift/testTekSwift/SFTableViewLayout.swift
//
//  SFTableViewLayout.swift
//  testTekSwift
//
//  Created by Nicolas Bellon on 29/10/2015.
//  Copyright © 2015 daid. All rights reserved.
//

import Foundation
import UIKit

let kAllConstraints : String = "kAllConstraints";
let kTopConstraint : String = "kTopConstraint";

class SFTableViewLayout : NSObject
{
    class func activateSFTableViewSuperViewConstraints(tableView: SFTableView)
    {
        guard let superView = tableView.superview
        else
        {
            return;
        }
        
        superView.translatesAutoresizingMaskIntoConstraints = false;
        
        var horizontalConstraints : [NSLayoutConstraint];
        var vertivalContstraints  :  [NSLayoutConstraint];
        
        var dest  :  [NSLayoutConstraint];
        
        let views : [String:UIView] = ["tableView": tableView];
        
        
       horizontalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("|-0-[tableView]-0-|", options: [], metrics: nil, views: views);
        
        vertivalContstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:|-0-[tableView]-0-|", options: [], metrics: nil, views: views);
        
        dest = horizontalConstraints + vertivalContstraints;
        
        NSLayoutConstraint.activateConstraints(dest);
    }
    
    class func activateScrollViewContraints(scrollView: UIScrollView)
    {
        
        guard scrollView.superview != nil
        else
        {
            return;
        }
        
        scrollView.translatesAutoresizingMaskIntoConstraints = false;
        scrollView.superview!.translatesAutoresizingMaskIntoConstraints = false;
        
        var horizontalConstraints : [NSLayoutConstraint];
        var vertivalContstraints  :  [NSLayoutConstraint];
        var equalWith : NSLayoutConstraint;
        var equalHeight : NSLayoutConstraint;
        
        var dest  :  [NSLayoutConstraint];
        
        let views : [String:UIView] = ["scrollView": scrollView];
        
        
        horizontalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("|-0-[scrollView]-0-|", options: [], metrics: nil, views: views);
        
        vertivalContstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:|-0-[scrollView]-0-|", options: [], metrics: nil, views: views);
        
        equalWith = NSLayoutConstraint.init(item: scrollView.superview!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: scrollView, attribute: NSLayoutAttribute.Width, multiplier: 1.0, constant: 0.0);
        
        equalHeight = NSLayoutConstraint.init(item: scrollView.superview!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: scrollView, attribute: NSLayoutAttribute.Height, multiplier: 1.0, constant: 0.0);
        
        
        dest = horizontalConstraints + vertivalContstraints;
        
        dest.append(equalWith);
        dest.append(equalHeight);
        
        NSLayoutConstraint.activateConstraints(dest);
    }

    class func activateInitialConstraintsForCell(cell: SFTableViewCell)
    {
        guard cell.superview != nil else
        {
            return;
        }
        
        cell.translatesAutoresizingMaskIntoConstraints = false;
        
        var horizontalConstraints : [NSLayoutConstraint];
        var verticalConstraints : [NSLayoutConstraint];
        var dest                : [NSLayoutConstraint];
        let views               : [String:UIView] = ["Cell":cell];
        let metrics             : [String:String] = ["cellHeight":SFTableViewCell.cellHeight().description, "cellWidth":SFTableViewCell.cellWidth().description];
        
        horizontalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("|-0-[Cell(==cellWidth)]-0-|", options: [], metrics: metrics, views: views);
        verticalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:[Cell(==cellHeight)]", options: [], metrics: metrics, views: views);
        
        dest = horizontalConstraints + verticalConstraints;
        
        NSLayoutConstraint.activateConstraints(dest);
        self.activateClassicModeForCell(cell);
    }
    
    class func activateClassicModeForCell(cell: SFTableViewCell)
    {
        var verticalConstraints : [NSLayoutConstraint];
        var lastCellConstraints : [NSLayoutConstraint] = [];
        let views           : [String: UIView];
        let metrics         : [String:String] = ["spaceBetweenCells":"10"];
        var dest            : [NSLayoutConstraint] = [];
        
        guard cell.superview != nil else
        {
            return;
        }
        
        cell.superview!.removeConstraints(cell.cellConstraint);
        cell.cellConstraint.removeAll();
        
        if cell.previousCell != nil
        {
            views = ["Cell":cell, "PreviousCell":cell.previousCell!];
            verticalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:[PreviousCell]-spaceBetweenCells-[Cell]", options: [], metrics: metrics, views: views)
        }
        else
        {
            views = ["Cell":cell];
            verticalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:|-0-[Cell]", options: [], metrics: nil, views: views)

        }
        dest = verticalConstraints;
        
        if cell.nextCell == nil
        {
            lastCellConstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:[Cell]-0-|", options: [], metrics: nil, views: views);
            dest.appendContentsOf(lastCellConstraints);
        }
        
        if (cell.topConstraint != nil)
        {
            cell.topConstraint = nil;
        }
        
        cell.topConstraint = verticalConstraints.last;
        
        NSLayoutConstraint.activateConstraints(dest);
        cell.cellConstraint = dest;
    }
    
    class func activateMoovingModeForCell(cell: SFTableViewCell)
    {
        guard cell.superview != nil else
        {
            return;
        }
        
        let views : [String:UIView] = ["Cell":cell];

        cell.superview!.removeConstraints(cell.cellConstraint);

        cell.cellConstraint.removeAll();
        
        let verticalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:|-0-[Cell]", options:[],metrics:nil,views:views);
        
        NSLayoutConstraint.activateConstraints(verticalConstraints);
        cell.topConstraint = verticalConstraints.last;
        cell.cellConstraint = verticalConstraints;
    }

}